ethers-solc/src/project_util.rs

```diff
@@ -0,0 +1,191 @@
+//! Utilities for mocking project workspaces
+use crate::{
+    cache::SOLIDITY_FILES_CACHE_FILENAME,
+    config::ProjectPathsConfigBuilder,
+    error::{Result, SolcError},
+    hh::HardhatArtifacts,
+    remappings::Remapping,
+    ArtifactOutput, MinimalCombinedArtifacts, Project, ProjectCompileOutput, ProjectPathsConfig,
+    SolcIoError,
+};
+use fs_extra::{dir, file};
+use std::path::Path;
+use tempdir::TempDir;
+
+pub struct TempProject<T: ArtifactOutput> {
+    /// temporary workspace root
+    root: TempDir,
+    /// actual project workspace with the `root` tempdir as its root
+    inner: Project<T>,
+}
+
+impl<T: ArtifactOutput> TempProject<T> {
+    /// Makes sure all resources are created
+    fn create_new(root: TempDir, inner: Project<T>) -> std::result::Result<Self, SolcIoError> {
+        let project = Self { root, inner };
+        project.paths().create_all()?;
+        Ok(project)
+    }
+
+    pub fn new(paths: ProjectPathsConfigBuilder) -> Result<Self> {
+        let tmp_dir = TempDir::new("root").map_err(|err| SolcError::io(err, "root"))?;
+        let paths = paths.build_with_root(tmp_dir.path());
+        let inner = Project::builder().artifacts().paths(paths).build()?;
+        Ok(Self::create_new(tmp_dir, inner)?)
+    }
+
+    pub fn project(&self) -> &Project<T> {
+        &self.inner
+    }
+
+    pub fn compile(&self) -> Result<ProjectCompileOutput<T>> {
+        self.project().compile()
+    }
+
+    pub fn project_mut(&mut self) -> &mut Project<T> {
+        &mut self.inner
+    }
+
+    /// The configured paths of the project
+    pub fn paths(&self) -> &ProjectPathsConfig {
+        &self.project().paths
+    }
+
+    /// The root path of the temporary workspace
+    pub fn root(&self) -> &Path {
+        self.project().paths.root.as_path()
+    }
+
+    /// Returns the handle to the tempdir and the project
+    ///
+    /// NOTE: the `TempDir` object deletes its directory on drop, also removing all the project's
+    /// content
+    pub fn split(self) -> (Project<T>, TempDir) {
+        (self.inner, self.root)
+    }
+
+    /// Copies a single file into the given dir
+    pub fn copy_file(&self, source: impl AsRef<Path>, target_dir: impl AsRef<Path>) -> Result<()> {
```

Comment:
Think this shouldn't be needed?

```suggestion
    pub fn copy_file(source: impl AsRef<Path>, target_dir: impl AsRef<Path>) -> Result<()> {
```